### PR TITLE
Add basic login

### PR DESCRIPTION
**Pull request by rogerhutchings**
Package: app-project

Adds modals for logging in, a user object store, and hooks up ZooHeader. It uses the existing panoptes auth client, and is added to the project app for now; eventually it'll get abstracted out.

## To do:

- [x] Tests
- [x] Add error feedback
- [ ] ~~Add register modal~~ - I'll do this in a separate PR

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

### CODE REVIEW

**Comment by eatyourgreens** [comment]
Login and logout works 👍 

However, the login modal isn't focused when it opens, which stops me from signing in if I can't use a mouse. I'm also able to tab out of the modal box so I'm not sure if it actually is modal.

The user account menu is styled with a serif font.
Screenshot showing the teal user menu using the wrong text colour and font.

**Comment by eatyourgreens** [comment]
Keyboard support for the user account menu has also vanished, relative to PFE. If you open your account menu on zooniverse.org, you can navigate the menu with the cursor keys and activate links with enter.

**Comment by rogerhutchings** [comment]
Cool, I'll add keyboard support in. Being able to tab outside the modal is a tricky one, since it uses using Grommet's `Layer`; we might have to write a wrapper component for that.

The font issue is with the shared components package, so I'm going to punt that for now.

**Comment by eatyourgreens** [comment]
There’s a wrapper in PFE, that I wrote, which takes care of basic focus handling in modals. Feel free to reuse it.

**Comment by eatyourgreens** [comment]
Did the user menu navigation get broken in the move to Grommet's Menu component? I can look at fixing that.

**Comment by rogerhutchings** [comment]
I think it's actually related to #327...

I've added a decorator component based on the PFE one - if it's functionally okay, I'll move it to the right place and write tests for everything.

**Comment by eatyourgreens** [comment]
Sorry, I meant keyboard navigation breaking in the user menu. I can look at fixing that.

Playing around a bit more, Grommet's Layer prevents focus from passing to the rest of the page but does allow focus to pass to the browser chrome, so it's sort of modal if you don't mind tabbing through the location bar to get back to the popup. 🤔 Might be behaviour that can be configured in Grommet.

**Comment by eatyourgreens** [comment]
I've pulled down the latest commits but I can't close the login box on ESC. I'm not sure what I'm doing wrong.

**Comment by rogerhutchings** [comment]
Ah right - yes, keyboard support in the user menu would have been lost when it was rewritten for Grommet.

I'm not sure what the issue is with esc - I've just tried it out in Chrome / Safari / Firefox. Might need to stop and start the dev server?

**Comment by eatyourgreens** [comment]
I've started and stopped the dev server a couple of times. Code is up-to-date too. It looks like the text inputs are swallowing the keyboard input, including the ESC key.

**Comment by rogerhutchings** [comment]
Aha - you're right, when an input is focused, it is failing to close. I'll add that other key listener in from the PFE wrapper in.

**Comment by rogerhutchings** [comment]
Bumping the Grommet version seems to have fixed the focus issue, but inputs are still swallowing the Esc keypress

**Comment by eatyourgreens** [comment]
`autoFocus=true` on the first menu item might solve the problem of it losing focus, if Grommet supports that.

Upgrading sounds like a good idea.

The username input on login should also have `autoFocus=true`, I think. Should we set `autocomplete=true` as well?

**Comment by srallen** [comment]
The TextInput: 
https://github.com/grommet/grommet/blob/308378cbfc9afb3ae2785576a236a304b531b2b1/src/js/components/TextInput/TextInput.js#L357-L387

They have it wrapped by `Keyboard` so, it should have the standard set they've defined. Also, any additional props we define does get passed on (they call the additional props `rest` in the code). If there's a question on behavior, I'm happy to pass along a question in their slack channel too.

**Comment by rogerhutchings** [comment]
Ah, not in this case - https://github.com/grommet/grommet/blob/308378cbfc9afb3ae2785576a236a304b531b2b1/src/js/components/TextInput/TextInput.js#L270

**Comment by srallen** [comment]
Good catch. I wonder why they did that. They're very responsible in their slack, so stuff like this I've asked about before and they either have a reason or respond with asking for an issue to be opened so they can improve it. I'll check in with them about that behavior.

**Comment by srallen** [comment]
Also, double checking the storybook, esc is only stopped if focus is on the text input. I guess they are assuming you are in the middle of typing? If you aren't focused on the inputs, esc will close the layer: https://storybook.grommet.io/?selectedKind=Layer&selectedStory=Form&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybook%2Fstories%2Fstories-panel

**Comment by rogerhutchings** [comment]
There's not a lot else to do in a login modal other than type into inputs tho

**Comment by rogerhutchings** [comment]
Okay, where we're at right now:

- Bumping Grommet has fixed the issue where you could tab through elements under the `Layer`.
- I've autofocused the first input using refs, since `TextInput` doesn't support `autoFocus`.
- Grommet is intentionally preventing `esc` from propagating and closing the Layer while an input is in focus; @srallen is asking on their Slack about that.
- We need a `forgot password` link
- We need some feedback for incorrect details
- We need tests

Anything I've missed?

**Comment by srallen** [comment]
I'm asking them about it, so I'll see what they say and share back.

**Comment by eatyourgreens** [comment]
If you accidentally hit Enter on the Sign-In button, you should definitely be able to cancel that with Esc.

**Comment by eatyourgreens** [comment]
Would submitting on Enter work if the username/password fields were wrapped in a form? Having to tab to the Login button then hit Enter is getting on my nerves, and will probably anger volunteers.

**Comment by rogerhutchings** [comment]
Bumping Grommet also provides access to their new `Form` component, so that works now 🎉

**Comment by eatyourgreens** [comment]
I've just tried it and I can log in by pressing Enter. 🎉 

But:
- My password is rendered as text now, not asterisks.
- I can still tab out of the logout form into the browser Chrome. Press Tab a couple of times from the username field and the focus just vanishes, with no visible indication as to where it's gone.

**Comment by eatyourgreens** [comment]
Oh wait, the focus management HOC got removed. I missed that. I think that's why tab navigation through the form seems to be broken.

**Comment by eatyourgreens** [comment]
The username and password labels have been unlinked from the corresponding text inputs and I can't see why, but it might have happened after the Grommet upgrade.

**Comment by srallen** [comment]
@rogerhutchings I think `TextInput` should support `autoFocus` since that is native browser behavior. Pretty much all the grommet comments will pass along any props you set on them. The docs just describe what's custom to Grommet, so they don't take the time to go over every prop that it can use if it's normal DOM stuff. I've forked their code sandbox and added `autoFocus` and it appears to work: https://codesandbox.io/s/y2pnqvzokx

I'm gonna take a closer look at some of the oddities with the styles. The menu absolutely should be getting the font from the theme, so something odd is going on here.

**Comment by srallen** [comment]
The Menu component definitely isn't getting the merged zooniverse theme, just the base Grommet theme. (The reason the background is the theme color is because that color is being directly set using a prop rather than set in the theme.) I don't understand why it isn't yet, but it works as expected in the react-components Storybook.

**Comment by eatyourgreens** [comment]
I'm having a look to see if I can figure out why the form labels have stopped working.

**Comment by eatyourgreens** [comment]
Oh wait, I should actually read the Grommet example:
> It typically contains an input control like TextInput, TextArea, Select, etc.

I think the code for the inputs is supposed to be something like:
```jsx
<FormField>
  <TextInput />
</FormField>
```

**Comment by rogerhutchings** [comment]
Yeah - the Grommet storybook suggests you can use a `FormField` alone, which defaults to a `TextInput`. Which is true, but I think the default accessibility is broken. I'll revert it back next.

**Comment by eatyourgreens** [comment]
The labels in the storybook work too, because the storybook code explicitly wraps a `TextInput` in a `FormField`.

I vaguely remember trying to debug problems like this when I was converting [zooniverse/pandora](https://github.com/zooniverse/pandora) over to Grommet. Their docs sometimes don't quite match actual usage in their example code.

**Comment by srallen** [comment]
The original PR where they added the esc event prevention doesn't really have a contextual explanation about why they made that change. I think I'd like to submit a PR to Grommet to make it configurable and see how they react. I asked about it in slack and didn't get a satisfactory response either since the dev who made the change hasn't been responsive recently.

**Comment by rogerhutchings** [comment]
This now has tests, and should be good to go - does it look okay @eatyourgreens?

**Comment by eatyourgreens** [comment]
I can't run the tests. I'm getting a module not found error for sinon.

File: packages/app-project/components/AuthModals/components/LoginModal/LoginModalContainer.spec.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+let wrapper
+let componentWrapper
+
+const fakeSignIn = sinon.fake.resolves()
+
+const AUTH_CLIENT = {
+  signIn: fakeSignIn
+}
+
+describe.only('Component > LoginModalContainer', function () {
```
Comment (eatyourgreens) [inline packages/app-project/components/AuthModals/components/LoginModal/LoginModalContainer.spec.js]: Remove this `.only`.

**Comment by eatyourgreens** [comment]
Running the app locally gives `Unhandled Rejection (TypeError): Cannot read property 'dark' of undefined` in the browser.

**Comment by rogerhutchings** [comment]
Looks like some changes to the components lib have broken the app. I'm looking into it

**Comment by rogerhutchings** [comment]
Doing a `lerna bootstrap` and rebuilding the components lib worked for me, try that.

**Comment by eatyourgreens** [comment]
I've got the app running on localhost:3000. Signing out works, but signing in fails and the modal displays `missing translation: en.LoginModal.errors.You need to sign in or sign up before continuing`

**Comment by rogerhutchings** [comment]
@eatyourgreens that should do it!

**Comment by eatyourgreens** [comment]
@rogerhutchings looks good to me. Have you looked at [`FormData`](https://developer.mozilla.org/en-US/docs/Web/API/FormData/FormData) for getting form state?

**Comment by eatyourgreens** [comment]
By the way, there's one failing test.
```
1) Component > LoginModalContainer
       onSubmit method
         should call the authClient signIn method:
     TypeError: Cannot read property 'querySelectorAll' of undefined
      at LoginModalContainer.querySelectorAll [as createLoginPayload] (components/AuthModals/components/LoginModal/LoginModalContainer.js:21:36)
      at LoginModalContainer.createLoginPayload (components/AuthModals/components/LoginModal/LoginModalContainer.js:31:31)
      at Context.onSubmit (components/AuthModals/components/LoginModal/LoginModalContainer.spec.js:44:26)
```

**Pull request merged**